### ISSUE TITLE
Should we support sync execution?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/testing#7]
Currently I started to implement the lib with naive approach. Something like having `Test` objects and call `test.execute(params)`. This is problematic, because we cannot be sure that the execution is sync.

**Comment by inf3rno** [comment]
- One solution to execute everything async.
- Another solution to have an `AsyncTest` class.
- Yet another solution  that the `Test` should not be responsible for its own execution.

**Comment by inf3rno** [comment]
In the last the `Test` is more like a data structure rather than an object. We could say that its responsibility is storing information for a single test. On the other hand we could say that its responsibility is executing a single test. In the latter one we should pass the event bus and the context too, not just the params.

**Comment by inf3rno** [comment]
I like this latter approach. I mean if we start with a test data structure, then all the groups will be data structures too. And all the execution logic will go to the executor, which will grow too big. Executing these tests looks completely reusable, so moving all the code to the executor does not make sense to me. Maybe later I figure out that this is an issue, but for now it isn't.

**Comment by inf3rno** [comment]
If we do it this way then the Group should be responsible for executing all the contained Tests. And there we will have a problem, because the Group will be responsible for parallel / series execution, dependency resolving, etc. which should be the responsibility of the Executor. At least the idea was that we can execute parallel simply by replacing the Executor and we don't need to touch the Tests at all.

**Comment by inf3rno** [comment]
Well in reality by parallel we need to give the dependencies, while by series the order can store them. We need to decide too whether they can use the same context or not. Normally I would create a new context for every test. At least that is the usual. But some tests depend on each other.

```js
{
    x: 1,
    test1(){
        assert.equal(this.x, 1);
        this.x++;
    },
    test2(){
        assert.equal(this.x, 2);
    }
}
```

It would be nice to have this kind of groups.

**Comment by inf3rno** [comment]
On the other hand for the sake of parallel execution we need to avoid series as much as possible. A simple solution to execute groups parallel and tests in series and whenever tests don't depend on each other we define a group. While this is a simple approach it is very inconvenient. We would end up with something like this:

```js
[
	{
		test1(x){
			assert.equal(x, 1);
		}
	},
	{
		test2(x){
			assert.equal(x, 2);
		}
	}
]
```

**Comment by inf3rno** [comment]
Either we need to give `execution: parallel/series` or `new ParallelGroup(o)` or make parallel default and give `dependencies: [[group.test2, group.test1],...]`.

**Comment by inf3rno** [comment]
I tend to agree with parallel by default and define dependencies. I am not entirely sure how to define dependencies. We don't necessarily need to do this in a config approach, something like adding a dependency array or a links array. While it would be useful for printing a dependency graph it is a lot more convenient for me to use code instead of configuration.

**Comment by inf3rno** [comment]
I am pretty sure tests should be responsible for their own execution too and group should be responsible for awaiting all its tests and make a group specific summary. We can do a plan with a root group too, I am just not sure what extra it would add. Probably it can contain and inject the context and have a plan title, description, etc. Or it could be a single entry point for reporters, etc. to attach to.

**Comment by inf3rno** [comment]
For the code of the Test class I am working on the important part is that I think I'll make every test async and the dependency management will be probably something like `await otherTest`. We can have concurrency issues when we use the same group instance with all of the tests. I mean

```js
{
    async a(){
        this.x = 1;
        this.y = await f();
        assert.equal(this.x, 1);
        assert.equal(this.y, 1);
    }
    async b(){
        this.x = 2;
        this.y = await g();
        assert.equal(this.x, 2);
        assert.equal(this.y, 2);
    }
}
```

Here the value of the `this.x` can change while we are waiting for the value of `this.y`. I don't want to expect the programmer to be smart and do the assertion before await. S/he has every right to expect the value not to change.

The usual solution in this grouping vs context scenario is creating a new instance for every test. Note that we need the context to be there, because we can add helper functions to it which can be reused by all our tests. Though we can use global function for that too or a separate helper module. As of communication between tests in this scenario ppl use return values or a separate context object. For me it is not acceptable, that we cannot use the properties of the actual context for communication. So there should be a better solution for this.

**Comment by inf3rno** [comment]
Well I guess we should add some rules to our system. If we want to use the context for communication we must be sure that people understand the consequences. Which means we document that there can be concurrency problems between tests if they use it inappropriately. Another solution is append only or raising events for communication. Yet another solution, which solves the dependency issue too is when tests can call each other, so they can directly communicate.

```js
{
    async a(){
        ...
        return 3;
    }
    async b(){
        var x = await this.x();
        ...
    }
}
```

We need to have very smart wrappers for it though and there will be another problem when we try to pass parameters from outside. I think we can call these outside parameters context or maybe even config and pass that config to the constructor of the group.

**Comment by inf3rno** [comment]
I don't like class extension here, it would be a lot better to have `new Group({...})` instead. For example:

```js
var g = new Group({
    async configure(p,q,r){
        this.p = p;
        this.q = q;
        this.r = r;
    }
    async a(){
        ...
        return 3;
    }
    async b(){
        var x = await this.x();
        ...
    }
});
```

**Comment by inf3rno** [comment]
While this is fine, we need to solve some problems with it. We don't know what is a test and what is a helper function. We need to document the special functions like configure.

We have 2 options here by distinguishing tests. One is adding some sort of code that helps distinguishing them, for example adding a test prefix or adding code to the group which registers tests one by one or based on a common trait like the already mentioned prefix. Another solution is using the Test class somehow to describe classes, but that would go against simplicity.

```js
var g = new Group({
    async configure(p,q,r){
        this.p = p;
        this.q = q;
        this.r = r;
    },
    a: new Test(async function(){
        ...
        return 3;
    })
    b: new Test(async function(){
        var x = await this.a();
        ...
    })
});
```

**Comment by inf3rno** [comment]
So from one point it is good to have a Test class, from another point an async function is just good enough. In theory. Having a separate class is good for adding description, label, helper functions, multiple steps, etc. too, but we can wait the same from groups too.

**Comment by inf3rno** [comment]
So I think we can get rid of the Test class. It did not do much anyways, it just stored the label and the implementation.

**Comment by inf3rno** [comment]
We can continue with the group.
```js
var g = new Group({
    async configure(p,q,r){
        this.p = p;
        this.q = q;
        this.r = r;
    }
    async a(){
        ...
        return 3;
    }
    async b(){
        var x = await this.a();
        ...
    }
});
```

**Comment by inf3rno** [comment]
I wonder if we really should call it group instead of test. I mean in our original approach groups shared the same interface as tests which means they had an execute method. Now that we don't have tests we still have two different type of groups. One contains these async functions we call tests, while the others contain groups only.

**Comment by inf3rno** [comment]
We could do something like

```js
var g = new Group({
    async configure(p,q,r){
        this.p = p;
        this.q = q;
        this.r = r;
    }
    async execute(){
        await Promise.all(this.a(), this.b());
    }
    async a(){
        ...
        return 3;
    }
    async b(){
        var x = await this.a();
        ...
    }
});
```

**Comment by inf3rno** [comment]
What makes the story rather complex is something like "should groups depend on each other?". I think they should. For example if we test ES6 features, then we should first test ES5 features, because they build on them and without proper ES5 support, ES6 cannot be supported. The same things can happen with test groups except if we have monster tests that are massive in size. I like the idea of conding groups very much, but I think the upper way it just bad. It would be a lot better to have classes from this perspective and if we have external parameters to configure them, then it would feel natural to make a new instance of the class with different parameters.

**Comment by inf3rno** [comment]
One problem to solve by execution is caching the return value. 

```js
    async execute(){
        await Promise.all(this.a(), this.b());
    }
    async a(){
        ...
        return 3;
    }
    async b(){
        var x = await this.a();
        ...
    }
```

Here we call test "a" twice. It would be nice to call it only once. Which means we need to wrap tests.

As of the external parameters instead of having a configure method we could simply do

```js
    async execute(p,q,r){
        await Promise.all(this.a(p,q,r), this.b(p,q,r));
    }
```

From one perspective it would be easier, from a different perspective it would be an issue, because if we call the same test twice with different parameters it can keep state from the previous run.

**Comment by inf3rno** [comment]
Maybe we should make tests similar to promises, which means they can either resolve or reject but they can do it only once. And if we want to run the same test again we need to make a new instance of it.

**Comment by inf3rno** [comment]
From one perspective this is good, but what about the methods inside the test? If we add a wrapper to it, then those can be called only once too. So if we want to benchmark for example, then those methods sucks, because each call needs a new test instance. Though we can move the repeating part to a helper function.

**Comment by inf3rno** [comment]
Well I redesigned the code with async tests, but in theory we can have sync test functions and use an asnyc execute function to run them. I don't think having async functions is a big overhead, so I'd rather not support sync execution.
